runtime error: invalid memory address or nil pointer dereference

### Nomad version
Nomad v1.7.2
BuildDate 2023-12-13T19:59:42Z
Revision 64e3dca9274b493e38a49fda3a70fd31d0485b91


### Operating system and Environment details
Amazon Linux 2 / Amazon Linux 2023

### Issue
```
2024-01-06T19:04:30.003Z [ERROR] http: http: panic serving NN.NN.NN.NNN:39804: runtime error: invalid memory address or nil pointer dereference                              
goroutine 3299 [running]:                                                                                                                                                    
net/http.(*conn).serve.func1()                                                                                                                                               
        net/http/server.go:1868 +0xb9                                                                                                                                        
panic({0x2a60280?, 0x4f15a40?})                                                                                                                                              
        runtime/panic.go:920 +0x270                                                                                                                                          
github.com/hashicorp/nomad/nomad/structs.(*Job).Validate(0xc005aefb00)                                                                                                       
        github.com/hashicorp/nomad/nomad/structs/structs.go:4677 +0x1c60                                                                                                     
github.com/hashicorp/nomad/nomad.(*jobValidate).Validate(0xc00084ef78, 0xc005aefb00)                                                                                         
        github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:356 +0x4e                                                                                                     
github.com/hashicorp/nomad/nomad.(*Job).admissionValidators(0xc000e36460, 0x40b0b8?)                                                                                         
        github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:138 +0x13c                                                                                                    
github.com/hashicorp/nomad/nomad.(*Job).admissionControllers(0xc0006da000?, 0x3038d73?)                                                                                      
        github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:105 +0x51                                                                                                     
github.com/hashicorp/nomad/nomad.(*Job).Plan(0xc000e36460, 0xc00617c300, 0xc00417e310)                                                                                       
        github.com/hashicorp/nomad/nomad/job_endpoint.go:1804 +0x196                                                                                                         
reflect.Value.call({0xc00009f680?, 0xc00084f160?, 0x7f61c50dad98?}, {0x303a316, 0x4}, {0xc005735538, 0x3, 0x0?})                                           
        reflect/value.go:596 +0xce7                                                   
reflect.Value.Call({0xc00009f680?, 0xc00084f160?, 0x4a7435?}, {0xc005735538?, 0xc005735588?, 0xa19bc7?})                                                                                                           
        reflect/value.go:380 +0xb9                                                                       
net/rpc.(*service).call(0xc000e266c0, 0x2db8640?, 0x9?, 0x0, 0xc000643880, 0x40?, {0x2f07880?, 0xc00617c300?, 0x0?}, {0x2bd9ce0, ...}, ...)                                                                        
        net/rpc/server.go:382 +0x211                                                                     
net/rpc.(*Server).ServeRequest(0x415325?, {0x37e2038, 0xc004ae6100})                                                                                                                                               
        net/rpc/server.go:503 +0x165                                                                     
github.com/hashicorp/nomad/nomad.(*Server).RPC(0xc0006da000, {0x304de73, 0x8}, {0x2f07880?, 0xc00617c2a0}, {0x2bd9ce0?, 0xc00417e2a0})                                                                             
        github.com/hashicorp/nomad/nomad/server.go:1984 +0xeb                                                          
github.com/hashicorp/nomad/command/agent.(*Agent).RPC(0xc005fd5420?, {0x304de73?, 0xc0049e9700?}, {0x2f07880?, 0xc00617c2a0?}, {0x2bd9ce0?, 0xc00417e2a0?})                  
        github.com/hashicorp/nomad/command/agent/agent.go:1274 +0x11b                                                                                                        
github.com/hashicorp/nomad/command/agent.(*HTTPServer).jobPlan(0xc005fd5420, {0x37de6f0, 0xc00617c120}, 0xc0049e9700, {0xc006adac0c, 0x15})                                                                        
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:189 +0x324                              
github.com/hashicorp/nomad/command/agent.(*HTTPServer).JobSpecificRequest(0x0?, {0x37de6f0, 0xc00617c120}, 0xc0049e9700)                                                                                                                      
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:84 +0x405                                             
github.com/hashicorp/nomad/command/agent.(*HTTPServer).registerHandlers.(*HTTPServer).wrap.func4({0x37de6f0, 0xc00617c120}, 0xc0049e9700)                                                                                                     
        github.com/hashicorp/nomad/command/agent/http.go:715 +0x168                                                    
net/http.HandlerFunc.ServeHTTP(0xc000105800?, {0x37de6f0?, 0xc00617c120?}, 0xc004ad99e8?)                                                                                                                                                     
        net/http/server.go:2136 +0x29               
net/http.(*ServeMux).ServeHTTP(0x0?, {0x37de6f0, 0xc00617c120}, 0xc0049e9700)                                          
        net/http/server.go:2514 +0x142                     
github.com/hashicorp/nomad/command/agent.NewHTTPServers.CompressHandler.CompressHandlerLevel.func3({0x37d6e70?, 0xc0007c0540}, 0xc0049e9700)                                                                                                  
        github.com/gorilla/handlers@v1.5.1/compress.go:141 +0x547                                                      
net/http.HandlerFunc.ServeHTTP(0x415325?, {0x37d6e70?, 0xc0007c0540?}, 0xc0007c0501?)                                  
        net/http/server.go:2136 +0x29                      
net/http.serverHandler.ServeHTTP({0x37ce378?}, {0x37d6e70?, 0xc0007c0540?}, 0x6?)                                      
        net/http/server.go:2938 +0x8e                      
net/http.(*conn).serve(0xc004429170, {0x37e1438, 0xc0067bfad0})                                                        
        net/http/server.go:2009 +0x5f4                     
created by net/http.(*Server).Serve in goroutine 347                                                                   
        net/http/server.go:3086 +0x5cb                     
2024-01-06T19:04:37.658Z [INFO]  nomad: setting up raft bolt store: no_freelist_sync=false       
```

### Reproduction steps
submitting a particular job via Terraform (1.6.6) and the Terraform provider (2.1.0) causes the server leader to crash.

#### Expected Result
This should not happen. Downgrading the servers to 1.6.5 make the error go away.

#### Actual Result
server crashes, leader changes

### Job file (if appropriate)

too big and customized to share for now.

### Nomad Server logs (if appropriate)

the segfault traceback has been posted above

### Nomad Client logs (if appropriate)
N/A

NOTE: This has started occurring recently.

I have already upgraded all the clients to 1.7.2

How long before I would have to downgrade all client to 1.6.5 as well? OR can I leave the cluster like this until 1.7.3 (hopefully with fixes)?

--- a/.changelog/19652.txt
+++ b/.changelog/19652.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+server: Fix panic when validating non-service reschedule block
+```

--- a/nomad/structs/structs.go
+++ b/nomad/structs/structs.go
@@ -4674,7 +4674,7 @@ func (j *Job) Validate() error {
 		}
 
 		if tg.MaxClientDisconnect != nil &&
-			tg.ReschedulePolicy.Attempts > 0 &&
+			(tg.ReschedulePolicy != nil && tg.ReschedulePolicy.Attempts > 0) &&
 			tg.PreventRescheduleOnLost {
 			err := fmt.Errorf("max_client_disconnect and prevent_reschedule_on_lost cannot be enabled when rechedule.attempts > 0")
 			mErr.Errors = append(mErr.Errors, err)

--- a/nomad/structs/structs_test.go
+++ b/nomad/structs/structs_test.go
@@ -470,6 +470,39 @@ func TestJob_ValidateNullChar(t *testing.T) {
 	assert.Error(job.Validate(), "null character in task name should not validate")
 }
 
+func TestJob_Validate_DisconnectRescheduleLost(t *testing.T) {
+	ci.Parallel(t)
+
+	// Craft our speciality jobspec to test this particular use-case.
+	testDisconnectRescheduleLostJob := &Job{
+		ID:     "gh19644",
+		Name:   "gh19644",
+		Region: "global",
+		Type:   JobTypeSystem,
+		TaskGroups: []*TaskGroup{
+			{
+				Name:                    "cache",
+				MaxClientDisconnect:     pointer.Of(1 * time.Hour),
+				PreventRescheduleOnLost: true,
+				Tasks: []*Task{
+					{
+						Name:   "redis",
+						Driver: "docker",
+						Config: map[string]interface{}{
+							"image": "redis:7",
+						},
+						LogConfig: DefaultLogConfig(),
+					},
+				},
+			},
+		},
+	}
+
+	testDisconnectRescheduleLostJob.Canonicalize()
+
+	must.NoError(t, testDisconnectRescheduleLostJob.Validate())
+}
+
 func TestJob_Warnings(t *testing.T) {
 	ci.Parallel(t)
 

